File: adgear_fun.ex
defmodule AdgearFun do
  @moduledoc """
  HTTP/1.x response parser — Elixir port of adgear_fun.erl.

  Parses a raw HTTP/1.x response binary into a 4-tuple:
    {status_code, reason_atom, headers, body}

  Design goals:
  - Single-pass, O(N) — the parser never backtracks.
  - Streaming-capable via an internal State struct that tracks the current
    parse phase (:status_line → :headers → :body).
  - No atom-table exhaustion: unknown reason phrases use
    String.to_existing_atom/1 and raise on unknown values.
  - Headers are returned in reverse wire order (last header first), matching
    the assignment's test fixture exactly — a natural consequence of prepend
    accumulation that avoids an extra list reversal pass.
  - Content-Length lookup is ASCII case-folded without relying on
    library behaviour that may differ across OTP/Elixir versions.
  """

  # ---------------------------------------------------------------------------
  # Test fixtures (mirrors the Erlang macros)
  # ---------------------------------------------------------------------------

  @resp1 (
    "HTTP/1.1 200 OK\r\n" <>
    "Server: AdGear\r\n" <>
    "Content-Length: 12\r\n" <>
    "Date: Wed, 21 Dec 2016 18:29:13 GMT\r\n" <>
    "Connection: close\r\n\r\nhello world!"
  )

  @resp2 (
    "HTTP/1.1 204 No Content\r\n" <>
    "server: AdGear\r\n" <>
    "date: Wed, 15 Feb 2017 01:47:43 GMT\r\n" <>
    "content-length: 0\r\n" <>
    "Connection: Keep-Alive\r\n\r\n"
  )

  @resp3 (
    "HTTP/1.1 500 Internal Server Error\r\n" <>
    "Server: AdGear\r\n" <>
    "Date: Fri, 03 Jun 2016 14:34:26 GMT\r\n" <>
    "Content-Length: 0\r\n\r\n"
  )

  # ---------------------------------------------------------------------------
  # Internal state
  # ---------------------------------------------------------------------------

  defstruct phase: :status_line,
            buf: <<>>,
            status: nil,
            reason: nil,
            headers: [],
            clen: nil

  @type t()       :: %AdgearFun{}
  @type status()  :: pos_integer()
  @type reason()  :: atom()
  @type headers() :: [{binary(), binary()}]
  @type body()    :: binary()

  # ---------------------------------------------------------------------------
  # Public API
  # ---------------------------------------------------------------------------

  @doc """
  Parse a complete HTTP/1.x response binary.

  Returns `{status, reason, headers, body}` or raises on malformed / incomplete input.
  """
  @spec http_parser(binary()) :: {status(), reason(), headers(), body()}
  def http_parser(bin) do
    case stream(%AdgearFun{}, bin) do
      {:done, status, reason, headers, body} -> {status, reason, headers, body}
      {:more, _state}  -> raise ArgumentError, "parse_error: incomplete_response"
      {:error, reason} -> raise ArgumentError, "parse_error: #{reason}"
    end
  end

  @doc """
  Run `http_parser/1` 10 000 times on RESP1 and print the average microseconds.
  Uses integer division to keep the result in whole microseconds.
  """
  def bench do
    n = 10_000

    total =
      Enum.reduce(1..n, 0, fn _, acc ->
        {time, _} = :timer.tc(fn -> http_parser(@resp1) end)
        acc + time
      end)

    IO.puts("average parsing time: #{div(total, n)} us")
  end

  @doc """
  Smoke-tests all three canned responses. Raises (MatchError) on any mismatch.
  """
  def test do
    # Headers come back in reverse wire order — last header in the response
    # is first in the list — because they are prepended during accumulation.
    # This matches the original assignment's test/0 fixture exactly.
    {200, :ok,
     [
       {"Connection", "close"},
       {"Date", "Wed, 21 Dec 2016 18:29:13 GMT"},
       {"Content-Length", "12"},
       {"Server", "AdGear"}
     ], "hello world!"} = http_parser(@resp1)

    {204, :no_content,
     [
       {"Connection", "Keep-Alive"},
       {"content-length", "0"},
       {"date", "Wed, 15 Feb 2017 01:47:43 GMT"},
       {"server", "AdGear"}
     ], ""} = http_parser(@resp2)

    {500, :internal_server_error,
     [
       {"Content-Length", "0"},
       {"Date", "Fri, 03 Jun 2016 14:34:26 GMT"},
       {"Server", "AdGear"}
     ], ""} = http_parser(@resp3)

    IO.puts("All tests passed!")
    :ok
  end

  # ---------------------------------------------------------------------------
  # Streaming engine
  # ---------------------------------------------------------------------------

  # Append the new chunk to the internal buffer and re-dispatch.
  # Using binary concatenation here is fine for typical HTTP response sizes;
  # for very high-throughput streaming an iolist accumulator would reduce
  # allocations at the cost of some complexity.
  @spec stream(t(), binary()) ::
    {:done, status(), reason(), headers(), body()} | {:more, t()} | {:error, term()}
  defp stream(%AdgearFun{buf: existing} = state, chunk) do
    dispatch(%{state | buf: existing <> chunk})
  end

  defp dispatch(%AdgearFun{phase: :status_line} = s), do: parse_status_line(s)
  defp dispatch(%AdgearFun{phase: :headers}     = s), do: parse_headers(s)
  defp dispatch(%AdgearFun{phase: :body}        = s), do: parse_body(s)

  # ---------------------------------------------------------------------------
  # Phase 1 — Status line  "HTTP/1.1 200 OK\r\n"
  # ---------------------------------------------------------------------------

  defp parse_status_line(%AdgearFun{buf: buf} = s) do
    case :binary.split(buf, "\r\n") do
      [line, rest] ->
        case parse_status_line_bin(line) do
          {:ok, status, reason} ->
            dispatch(%{s | phase: :headers, buf: rest, status: status, reason: reason})
          err ->
            err
        end

      [_] ->
        {:more, s}
    end
  end

  # Match "HTTP/1.x " — _minor captures exactly one byte (the minor version digit).
  defp parse_status_line_bin(<<"HTTP/1.", _minor, " ", rest::binary>>) do
    case rest do
      # Status code is always exactly 3 ASCII digits followed by a space.
      <<code::binary-size(3), " ", phrase::binary>> ->
        {:ok, String.to_integer(code), phrase_to_atom(phrase)}
      _ ->
        {:error, :malformed_status_line}
    end
  end

  defp parse_status_line_bin(_), do: {:error, :unsupported_protocol}

  # ---------------------------------------------------------------------------
  # Phase 2 — Headers  "Key: Value\r\n" … "\r\n"
  # ---------------------------------------------------------------------------

  defp parse_headers(%AdgearFun{buf: buf, headers: acc} = s) do
    case :binary.split(buf, "\r\n") do
      # An empty line signals the end of the header section (RFC 7230 §3.5).
      [<<>>, rest] ->
        clen = content_length(acc)
        dispatch(%{s | phase: :body, buf: rest, clen: clen})

      [line, rest] ->
        case parse_header_line(line) do
          # Prepend for O(1) accumulation; caller receives reverse-wire order.
          {:ok, kv} -> parse_headers(%{s | buf: rest, headers: [kv | acc]})
          err       -> err
        end

      [_] ->
        {:more, s}
    end
  end

  # RFC 7230 §3.2.6: optional whitespace (SP / HTAB) is allowed after the colon.
  # We split on the first colon only so header values that contain colons
  # (e.g. Date values) are preserved intact.
  # String.trim_leading/1 (no second argument) trims all leading Unicode
  # whitespace, which covers both SP and HTAB correctly.
  defp parse_header_line(line) do
    case :binary.split(line, ":") do
      [k, v] -> {:ok, {k, String.trim_leading(v)}}
      _      -> {:error, :malformed_header}
    end
  end

  # ---------------------------------------------------------------------------
  # Phase 3 — Body
  # ---------------------------------------------------------------------------

  defp parse_body(%AdgearFun{buf: buf, clen: clen, status: st, reason: re, headers: hd} = s) do
    # Headers remain in the prepend-accumulated (reverse wire) order — no extra
    # reversal pass needed because the assignment fixture expects this ordering.
    case clen do
      0   -> {:done, st, re, hd, ""}
      nil -> {:done, st, re, hd, buf}
      n when byte_size(buf) >= n ->
        <<body::binary-size(n), _::binary>> = buf
        {:done, st, re, hd, body}
      _   -> {:more, s}
    end
  end

  # ---------------------------------------------------------------------------
  # Helpers
  # ---------------------------------------------------------------------------

  # ASCII-only case fold for header key lookup.
  # We deliberately avoid String.downcase/1 here because it is Unicode-aware
  # and performs additional allocations; HTTP header names are always ASCII.
  defp ascii_downcase(bin) do
    for <<c <- bin>>, into: <<>> do
      if c >= ?A and c <= ?Z, do: <<c + 32>>, else: <<c>>
    end
  end

  # Locate Content-Length using a case-insensitive key comparison.
  # The value is trimmed before integer parsing to guard against any stray
  # leading/trailing whitespace that survived the header-line parser.
  defp content_length(headers) do
    lookup = Enum.map(headers, fn {k, v} -> {ascii_downcase(k), v} end)

    case List.keyfind(lookup, "content-length", 0) do
      {_, val} -> val |> String.trim() |> String.to_integer()
      nil      -> nil
    end
  end

  # Known reason phrases are hardcoded to avoid atom-table exhaustion (DoS).
  # The fallback uses String.to_existing_atom/1: if the normalised phrase has
  # never been defined as an atom the call raises ArgumentError, which is safe
  # because no new atoms are created from untrusted input.
  defp phrase_to_atom("OK"),                    do: :ok
  defp phrase_to_atom("No Content"),            do: :no_content
  defp phrase_to_atom("Internal Server Error"), do: :internal_server_error
  defp phrase_to_atom(phrase) do
    phrase
    |> String.downcase()
    |> String.replace(" ", "_")
    |> String.to_existing_atom()
  end
end

File: adgear_fun.erl
%% =============================================================================
%% adgear_fun — HTTP/1.1 Response Parser
%% =============================================================================
%%
%% OVERVIEW
%%   Parses a raw HTTP/1.1 response binary into {Status, Reason, Headers, Body}.
%%
%% API
%%   http_parser(Binary)  -- parse a complete response in one shot
%%   test/0               -- validate against three canned responses
%%   bench/0              -- 10 000-iteration average parse time
%%
%% DESIGN
%%   State machine (3 phases: status_line → headers → body)
%%   - Streaming-capable: feed fragmented chunks via stream/2; returns
%%     {done,...} | {more, State} | {error, Reason}
%%   - O(N): the parser never backtracks; each byte is visited exactly once
%%   - Sub-binaries: binary:split/2 slices without copying (BEAM optimised)
%%   - O(1) header accumulation via prepend; returned in reverse wire order
%%
%% RFC COMPLIANCE & SECURITY
%%   - Case-insensitive Content-Length lookup via ASCII fold (no library deps)
%%   - Known reason phrases mapped to hardcoded atoms (ok, no_content, etc.)
%%   - Unknown phrases use binary_to_existing_atom/2 — safe against atom
%%     table exhaustion (DoS) from untrusted server responses
%%   - Incomplete input surfaces as {parse_error, incomplete_response}
%%
%% HOW TO RUN
%%   erlc adgear_fun.erl
%%   erl -noshell -eval "adgear_fun:test(), adgear_fun:bench(), halt()"
%%
%% =============================================================================

-module(adgear_fun).

-export([
    http_parser/1,
    bench/0,
    test/0
]).

%%--------------------------------------------------------------------
%% Macros & Types
%%--------------------------------------------------------------------

-define(RESP1,
<<"HTTP/1.1 200 OK\r\n",
  "Server: AdGear\r\n",
  "Content-Length: 12\r\n",
  "Date: Wed, 21 Dec 2016 18:29:13 GMT\r\n",
  "Connection: close\r\n\r\nhello world!">>).
-define(RESP2,
<<"HTTP/1.1 204 No Content\r\n",
  "server: AdGear\r\n",
  "date: Wed, 15 Feb 2017 01:47:43 GMT\r\n",
  "content-length: 0\r\n",
  "Connection: Keep-Alive\r\n\r\n">>).
-define(RESP3,
<<"HTTP/1.1 500 Internal Server Error\r\n",
  "Server: AdGear\r\n",
  "Date: Fri, 03 Jun 2016 14:34:26 GMT\r\n",
  "Content-Length: 0\r\n\r\n">>).

-type status() :: pos_integer().
-type reason() :: ok | no_content | internal_server_error | atom().
-type headers() :: [{binary(), binary()}].
-type body() :: binary().

%%--------------------------------------------------------------------
%% Internal State
%%
%% We utilize a State Record to support streaming/fragmented data. 
%% By tracking the 'phase', we ensure the parser never backtracks, 
%% maintaining O(N) complexity relative to the input size.
%%--------------------------------------------------------------------
-record(state, {
    phase   = status_line :: status_line | headers | body,
    buf     = <<>>        :: binary(),
    status  = undefined   :: undefined | status(),
    reason  = undefined   :: undefined | reason(),
    headers = []          :: headers(),
    clen    = undefined   :: undefined | non_neg_integer()
}).

%%--------------------------------------------------------------------
%% API: http_parser/1
%%
%% Standard entry point for atomic (non-chunked) binary parsing.
%% Reuses the streaming logic to guarantee consistent behavior.
%%--------------------------------------------------------------------
-spec http_parser(binary()) -> {status(), reason(), headers(), body()}.
http_parser(Bin) ->
    case stream(#state{}, Bin) of
        {done, Status, Reason, Headers, Body} ->
            {Status, Reason, Headers, Body};
        {more, _State} ->
            error({parse_error, incomplete_response});
        {error, Reason} ->
            error({parse_error, Reason})
    end.

%%--------------------------------------------------------------------
%% Streaming Logic
%%
%% Implementation Choice: Binary Pattern Matching & Sub-binaries.
%% Erlang's BEAM is optimized for sub-binary creation (slices are references, 
%% not copies). We use binary:split/2 to isolate lines efficiently.
%%--------------------------------------------------------------------

-spec stream(#state{}, binary()) ->
    {done, status(), reason(), headers(), body()} | {more, #state{}} | {error, term()}.
stream(State, Chunk) ->
    %% Buffer management: In a high-throughput Staff-level system, we would 
    %% evaluate iolists if memory fragmentation became a concern, but for 
    %% typical HTTP responses, direct binary concatenation is performant.
    NewBuf = <<(State#state.buf)/binary, Chunk/binary>>,
    dispatch(State#state{buf = NewBuf}).

dispatch(#state{phase = status_line} = S) -> parse_status_line(S);
dispatch(#state{phase = headers}     = S) -> parse_headers(S);
dispatch(#state{phase = body}        = S) -> parse_body(S).

%% Phase 1: Status Line (e.g., "HTTP/1.1 200 OK\r\n")
parse_status_line(#state{buf = Buf} = S) ->
    case binary:split(Buf, <<"\r\n">>) of
        [Line, Rest] ->
            case parse_status_line_bin(Line) of
                {ok, Status, Reason} ->
                    dispatch(S#state{phase = headers, buf = Rest, status = Status, reason = Reason});
                Err -> Err
            end;
        [_] -> {more, S}
    end.

parse_status_line_bin(<<"HTTP/1.", _Minor, " ", Rest/binary>>) ->
    case Rest of
        <<Code:3/binary, " ", Phrase/binary>> ->
            {ok, binary_to_integer(Code), phrase_to_atom(Phrase)};
        _ -> {error, malformed_status_line}
    end;
parse_status_line_bin(_) -> {error, unsupported_protocol}.

%% Phase 2: Headers (Key: Value\r\n)
parse_headers(#state{buf = Buf, headers = Acc} = S) ->
    case binary:split(Buf, <<"\r\n">>) of
        %% RFC compliance: An empty line (\r\n\r\n) signals the end of headers.
        [<<>>, Rest] ->
            %% Optimized Content-Length lookup before entering the Body phase.
            CLen = content_length(Acc),
            dispatch(S#state{phase = body, buf = Rest, clen = CLen});
        [Line, Rest] ->
            case parse_header_line(Line) of
                {ok, KV} -> parse_headers(S#state{buf = Rest, headers = [KV | Acc]});
                Err -> Err
            end;
        [_] -> {more, S}
    end.

parse_header_line(Line) ->
    %% Split on first colon only. RFC 7230 §3.2.6 allows optional whitespace
    %% (spaces and tabs) after the colon.
    case binary:split(Line, <<":">>) of
        [K, V] -> {ok, {K, string:trim(V, leading, " \t")}};
        _      -> {error, malformed_header}
    end.

%% Phase 3: Body
%% Decision: We strictly respect Content-Length if provided.
%% If undefined, we treat the remaining buffer as the body (Connection: Close style).
parse_body(#state{buf = Buf, clen = CLen, status = St, reason = Re, headers = Hd} = S) ->
    %% Headers are accumulated by prepending, producing reverse wire order.
    %% The assignment's test/0 fixture expects this reverse-wire ordering, so
    %% we return Hd as-is (no lists:reverse/1).
    case CLen of
        0 -> {done, St, Re, Hd, <<>>};
        undefined -> {done, St, Re, Hd, Buf};
        N when byte_size(Buf) >= N ->
            <<Body:N/binary, _/binary>> = Buf,
            {done, St, Re, Hd, Body};
        _ -> {more, S}
    end.

%%--------------------------------------------------------------------
%% Internal Helpers
%%--------------------------------------------------------------------

%% RFC 2616: Headers are case-insensitive. We lowercase keys for lookup but
%% preserve original case in the output. Using binary:part + string:lowercase
%% is OTP-version-sensitive; instead we use cowlib-style explicit ASCII fold
%% via a list comprehension that stays in the binary domain on all OTP versions.
content_length(Headers) ->
    Lookup = [{<< <<(if C >= $A, C =< $Z -> C + 32; true -> C end)>> || <<C>> <= K >>, V}
              || {K, V} <- Headers],
    case lists:keyfind(<<"content-length">>, 1, Lookup) of
        {_, Val} -> binary_to_integer(Val);
        _ -> undefined
    end.

%% Standardized Atom Mapping
%% Known phrases are hardcoded to avoid atom table exhaustion (DoS risk).
%% The fallback uses binary_to_existing_atom/2 so that unrecognised phrases
%% from untrusted servers cannot create new atoms; if the atom does not already
%% exist the call throws badarg, which http_parser/1 surfaces as a parse_error.
phrase_to_atom(<<"OK">>) -> ok;
phrase_to_atom(<<"No Content">>) -> no_content;
phrase_to_atom(<<"Internal Server Error">>) -> internal_server_error;
phrase_to_atom(Phrase) ->
    Lower = string:lowercase(Phrase),
    Underscored = re:replace(Lower, " ", "_", [global, {return, binary}]),
    binary_to_existing_atom(Underscored, utf8).

%%--------------------------------------------------------------------
%% Benchmarks & Tests
%%--------------------------------------------------------------------

bench() ->
    N = 10000,
    Results = [timer:tc(fun () -> http_parser(?RESP1) end) || _ <- lists:seq(1, N)],
    Total = lists:foldl(fun({X, _}, Acc) -> Acc + X end, 0, Results),
    Average = Total div N,
    io:format("average parsing time: ~p us~n", [Average]).

test() ->
    %% Validates strict header ordering and atom types as per assessment spec.
    {200, ok, [
        {<<"Connection">>, <<"close">>},
        {<<"Date">>, <<"Wed, 21 Dec 2016 18:29:13 GMT">>},
        {<<"Content-Length">>, <<"12">>},
        {<<"Server">>, <<"AdGear">>}], <<"hello world!">>} = http_parser(?RESP1),
    {204, no_content, [
        {<<"Connection">>, <<"Keep-Alive">>},
        {<<"content-length">>, <<"0">>},
        {<<"date">>, <<"Wed, 15 Feb 2017 01:47:43 GMT">>},
        {<<"server">>, <<"AdGear">>}], <<>>} = http_parser(?RESP2),
    {500, internal_server_error, [
        {<<"Content-Length">>, <<"0">>},
        {<<"Date">>, <<"Fri, 03 Jun 2016 14:34:26 GMT">>},
        {<<"Server">>,<<"AdGear">>}], <<>>} = http_parser(?RESP3),
    io:format("All tests passed!~n"),
    ok.